topydo/lib/ChangeSet.py

```diff
@@ -0,0 +1,133 @@
+# Topydo - A todo.txt client written in Python.
+# Copyright (C) 2014 - 2015 Bram Schoenmakers <me@bramschoenmakers.nl>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+""" This module serves for managing todo and archive changesets. """
+
+from os import path
+from topydo.lib.Config import config
+from hashlib import sha1
+import shelve
+from copy import deepcopy
+
+def get_backup_path():
+    dirname, filename = path.split(path.splitext(config().todotxt())[0])
+    filename = '.' + filename + '.bak'
+
+    return path.join(dirname, filename)
+
+class ChangeSet(object):
+    def __init__(self, p_todolist=None, p_archive=None):
+        self.todolist = deepcopy(p_todolist)
+        self.todolist_hash = self._hash_todolist(p_todolist)
+        self.archive = deepcopy(p_archive)
+        self.archive_hash = self._hash_todolist(p_archive)
+        self.shelve_file = shelve.open(get_backup_path())
+
+    def add_archive(self, p_archive):
+        """ Sets deep copy of p_archive as archive attribute and calculate archive_hash. """
+        self.archive = deepcopy(p_archive)
+        self.archive_hash = self._hash_todolist(p_archive)
+
+    def add_todolist(self, p_todolist):
+        """ Sets deep copy of p_todolist as todolist attribute and calculate todolist_hash. """
+        self.todolist = deepcopy(p_todolist)
+        self.todolist_hash = self._hash_todolist(p_todolist)
+
+    def save(self, p_todolist):
+        """
+        Saves a tuple with archive, todolist and their hashes into the
+        backup file. Tuple is indexed in backup file  with hash calculated from
+        p_todolist.
+        """
+        self.trim()
+
+        current_hash = self._hash_todolist(p_todolist)
+        self.shelve_file[current_hash] = (self.todolist, self.todolist_hash, self.archive, self.archive_hash)
+
+        index = self.get_index()
+        index.insert(0, current_hash)
+        self.save_index(index)
+
+        self.close()
+
+    def delete(self, p_todolist_hash):
+        """ Removes backup indexed with p_todolist_hash from backup file. """
+        index = self.get_index()
+
+        try:
+            del self.shelve_file[p_todolist_hash]
+            index.remove(p_todolist_hash)
+        except KeyError:
+            pass
+
+        self.save_index(index)
+
+    def get_index(self):
```

Comment:
Shouldn't this be an internal function? I.e.: `_get_index`. Same for `save_index` below.